Allow SDFs to accept a start state.

chinmaychandak opened the pull request:
This PR attempts to close one aspect of #335. To checkpoint the state of an SDF, we need to do these 3 things:
1. When a job starts, retrieve (either from Dask datasets/S3/disk) the last known SDF state from previous runs. This can just be a one-time function call to a UDF outside of streamz.
2. Pass this retrieved state into an SDF so that the new run of this streamz job can pick up from where it left off — basically, allow SDFs to accept an initial state.
3. Publish (either to Dask/S3/disk) the updated SDF at the end of every batch. 

This PR attempts to complete step 2 above. Steps 1 and 3 are yet to be implemented, and a proposal will follow up.

For this PR, I have added the capability for _most_ critical SDF operations (reductions, groupby, rolling, window, windowed-groupby) to accept a start state. This allows streamz pipelines with an SDF to pick up from an initial state. 

The implementation is as follows:
1. For reductions and groupby-aggs like `sum`, the result of the current batch itself is the start state for the next batch. 
2. For operations like `groupby.mean`, rolling, windowing, streamz maintains a state OR a historical backlog of events for accurate computations. In such cases, one needs to `emit` the entire `state` maintained along with the result at the end of every batch. This state can then be the start state for the next batch in case the job crashes. 

I had initially added the start parameter at the root level of the operation. For e.g., `sdf.groupby([cols], start=None)`, but then this will not work for groupby aggregate use cases, which are pretty common in pipelines. So, to account for cases like #303, I moved to an implementation like `sdf.groupby([cols]).sum(start=None)`.  **Please refer to test cases.**

I would highly appreciate an in-depth review and suggestions so that I can make this more efficient. Once the code review is complete, I can add documentation.

This PR also has some changes to the current documentation indentation, and a small fix for cudf compatibility with timedelta.

chinmaychandak commented:
All tests are passing, except one: `from_kafka`, which is unrelated to the PR. This might be due to an upstream change. @martindurant, @CJ-Wright, @jsmaupin any ideas?

chinmaychandak commented:
I'm trying to figure out why the Kafka test is failing. 

I ran the command: 
`STREAMZ_LAUNCH_KAFKA=true python -m pytest -vs streamz/tests/test_kafka.py::test_from_kafka`. 

I see a segfault with this trace: 
```
======================================== test session starts ========================================
platform darwin -- Python 3.7.1, pytest-5.4.3, py-1.7.0, pluggy-0.13.1 -- /Users/cchandak/anaconda3/bin/python3
cachedir: .pytest_cache
rootdir: /Users/cchandak/Desktop/streamz
plugins: doctestplus-0.2.0, arraydiff-0.3, remotedata-0.3.1, openfiles-0.3.1
collected 1 item

streamz/tests/test_kafka.py::test_from_kafka Finding streamz-kafka ...
Stopping eca205cbd215 ...
eca205cbd215
docker run -d -p 2181:2181 -p 9092:9092 --env ADVERTISED_HOST=127.0.0.1 --env ADVERTISED_PORT=9092 --name streamz-kafka spotify/kafka
Fatal Python error: Segmentation fault

Current thread 0x000070000a2dd000 (most recent call first):
  File "/Users/cchandak/Desktop/streamz/streamz/sources.py", line 416 in do_poll
  File "/Users/cchandak/Desktop/streamz/streamz/sources.py", line 423 in poll_kafka
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 1147 in run
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 1233 in inner
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/tornado/stack_context.py", line 300 in null_wrapper
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/tornado/ioloop.py", line 758 in _run_callback
  File "/Users/cchandak/anaconda3/lib/python3.7/asyncio/events.py", line 88 in _run
  File "/Users/cchandak/anaconda3/lib/python3.7/asyncio/base_events.py", line 1764 in _run_once
  File "/Users/cchandak/anaconda3/lib/python3.7/asyncio/base_events.py", line 528 in run_forever
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/tornado/platform/asyncio.py", line 132 in start
  File "/Users/cchandak/anaconda3/lib/python3.7/threading.py", line 865 in run
  File "/Users/cchandak/anaconda3/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/Users/cchandak/anaconda3/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x0000700009bce000 (most recent call first):
  File "/Users/cchandak/anaconda3/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
  File "/Users/cchandak/anaconda3/lib/python3.7/threading.py", line 865 in run
  File "/Users/cchandak/anaconda3/lib/python3.7/threading.py", line 917 in _bootstrap_inner
  File "/Users/cchandak/anaconda3/lib/python3.7/threading.py", line 885 in _bootstrap

Thread 0x000000011e0915c0 (most recent call first):
  File "/Users/cchandak/anaconda3/lib/python3.7/selectors.py", line 558 in select
  File "/Users/cchandak/anaconda3/lib/python3.7/asyncio/base_events.py", line 1728 in _run_once
  File "/Users/cchandak/anaconda3/lib/python3.7/asyncio/base_events.py", line 528 in run_forever
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/tornado/platform/asyncio.py", line 132 in start
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/tornado/ioloop.py", line 571 in run_sync
  File "/Users/cchandak/Desktop/streamz/streamz/utils_test.py", line 69 in test_func
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/python.py", line 182 in pytest_pyfunc_call
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/python.py", line 1477 in runtest
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/runner.py", line 135 in pytest_runtest_call
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/runner.py", line 217 in <lambda>
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/runner.py", line 244 in from_call
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/runner.py", line 217 in call_runtest_hook
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/runner.py", line 186 in call_and_report
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/runner.py", line 100 in runtestprotocol
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/runner.py", line 85 in pytest_runtest_protocol
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/main.py", line 272 in pytest_runtestloop
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/main.py", line 247 in _main
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/main.py", line 191 in wrap_session
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/main.py", line 240 in pytest_cmdline_main
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/callers.py", line 187 in _multicall
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 87 in <lambda>
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/manager.py", line 93 in _hookexec
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pluggy/hooks.py", line 286 in __call__
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/_pytest/config/__init__.py", line 125 in main
  File "/Users/cchandak/anaconda3/lib/python3.7/site-packages/pytest/__main__.py", line 7 in <module>
  File "/Users/cchandak/anaconda3/lib/python3.7/runpy.py", line 85 in _run_code
  File "/Users/cchandak/anaconda3/lib/python3.7/runpy.py", line 193 in _run_module_as_main
[2]    4353 segmentation fault  STREAMZ_LAUNCH_KAFKA=true python3 -m pytest -vs
```
Not sure what is happening. Can someone please help?

chinmaychandak commented:
Okay, all tests are now passing. Turns out the `from_kafka` test failure was a one-off. :( 

I can add more tests if needed and some documentation after the code is reviewed. Please do have a look.

chinmaychandak commented:
Random tests like `test_tcp` are failing now, can one of the maintainers please retrigger the travis CI build?

martindurant commented:
Agree about the flaky tests, but it is rather difficult - typically the CI test machines run slower and with unexpected latency compared to testing locally. We don't want to add sleep statements all over the place, although tolerances on waiting would be fine.

martindurant commented:
@CJ-Wright , there is quite a lot of good looking work here, do you think you'll have the chance to review?

@chinmaychandak , I didn't immediately spot doc or example or how the sdl_checkpoint parameter should be used. Are there other accumulation operations aside from SDFs that might benefit from setting not just an initial value, but also an initial state?

chinmaychandak commented:
> Agree about the flaky tests, but it is rather difficult - typically the CI test machines run slower and with unexpected latency compared to testing locally. We don't want to add sleep statements all over the place, although tolerances on waiting would be fine.

Should we just re-trigger the build in these cases then?

> there is quite a lot of good looking work here, do you think you'll have the chance to review?

I would really appreciate that, @CJ-Wright. Thanks @martindurant! 

> I didn't immediately spot doc or example or how the `sdf_checkpoint` parameter should be used. 

Apologies for not adding doc strings until now. I was hoping to get an initial review before doing so. To answer your question, `sdf_checkpoint` is a boolean which defaults to `False`. To ensure checkpointing in operations on SDFs _that need to return current state along with the result_, one needs to set this parameter to `True`. This will let the user know that they should expect the SDF operation's output to be `(checkpoint_state, result)`. How one handles this checkpoint state downstream is a different problem altogether. Jarod and I will work on tying this to a Kafka checkpoint as part of the metadata. Note that this is in addition to the `start` parameter which is just the initial state.

> Are there other accumulation operations aside from SDFs that might benefit from setting not just an initial value, but also an initial state?

I can't think of any other operations at this time. The initial state can be a complex Python object (dataframe, deque of dataframes, etc., as can be seen in the tests). This would primarily be obtained from an existing stream with SDFs when it crashes, so that when the stream restarts, it can be fed this last state. Basically, this is beneficial to streams that maintain state across batches — and SDFs are the only way to do that. If the stream has only within-each-batch operations, the Kafka checkpoint should take care of the crash already, and using an initial state along with this might be overkill. I cannot think of other accumulation operations without SDFs that can benefit from this initial state; even if they can, retrieving/getting an initial state in the first place would be inconvenient.

martindurant commented:
This sounds like a `with_state` parameter then, i.e., that the output is both the internal state and the calculated value. That doesn't sound like it should be specific to SDFs, though I can see why it's particularly useful for this case.

chinmaychandak commented:
> This sounds like a `with_state` parameter then

I agree. I was thinking of a better name than `sdf_checkpoint` myself. I'll change the name to  `with_state` then.

martindurant commented:
> Should we just re-trigger the build in these cases then?

It's certainly the workable, if annoying, option. Some projects have used a `@flaky` marker, meaning that either the test will be re-tried some times, or that a certain number of tests marked like this are allowed to fail on a given run.

CJ-Wright commented on streamz/core.py:
```diff
@@ -867,12 +867,16 @@ def __init__(self, upstream, func, start=no_default, returns_state=False,
             self.state = x
-            return self._emit(x, metadata=metadata)
+            if self.sdf_checkpoint:
+                return self._emit((self.state, x), metadata=metadata)
```
I've read through the above comments so maybe I have a better idea of what is going on now.

I agree with @martindurant that `sdf_checkpoint` might be a bit too on the nose as names go, since the output state could be used for any number of reasons beyond checkpointing.

My issue here is that the change of that flag causes the graph topology to be different, causing the pipeline to need to be rewritten. I can't think of any current node types that have this kind of behavior.

I don't have a great work around for this though. 

Maybe we could have it named something else so that it is clear that these two modes aren't interchangeable, but this could cause an explosion in the number of classes we track. 

You could add a callback that is called before every emit that takes the current state, but that completely removes it from the graph preventing it from accessing the event loop integration or any other graph based tooling.

You could write a specialty node that when emitted to by the accumulate node access that node's status and emits that. This could maybe be done with a rather odd looking map.
```python
s = Stream()
a = s.accululate(...)
out = a.map(...)...
to_state_store = a.map(lambda x: a.state)
to_state_store.sink(store)
```
As an independent class
```python
class StateGetter(Stream):
    def __init__(self, upstream):
        # XXX: check that upstream has a state if not fail
        self.upstream = upstream
        ...
    def update(self, x):
        return upstream.state
```
This might be one of the more promising approaches, it retains the original topology of the graph, the state storage is on the graph's event loop, and the class version of this even can fail at init time if used improperly.

chinmaychandak replied:
> `sdf_checkpoint` might be a bit too on the nose as names go

Yes, agreed. Will change the name to `with_state`. 

> since the output state could be used for any number of reasons beyond checkpointing.

Yes, that's true. But I haven't seen any streaming stacks to save internal state (which ends up being really complex) for any other reason than checkpointing. Just curious to understand what other cases we might need to save the state for. 

> I don't have a great work around for this though. 

I completely agree. But, as I mentioned above, this PR is targeted primarily for SDF operations to accept an initial state (step 2 in the PR description). @jsmaupin and I will be working on step 3 (publishing the updated SDF at the end of every batch/update). 

This step would handle the tuple of `(state, result)` emitted by the `accumulate` node,  or the `result` of some of the SDF group-by aggs (which is same as the state in these cases). So far, the idea I had to handle this was similar to yours: we add a `publish` node which the SDF aggregation node emits to. The `publish` node plucks the `state` out to the `metadata` already being used for checkpointing, and continues emitting only the `result` downstream. When the checkpointing mechanism is triggered to commit Kafka offsets, we add another callback to checkpoint/save this SDF state either on disk/S3/Dask. 

But, until this gets implemented, I thought we let this be as is, and let the downstream take care of how the state is handled. Hence, the `with_state` parameter is orthogonal to the `start` parameter — one might want to start the job with an SDF state, but might not need the state downstream anymore. But in order to get an initial state (which is usually a complex object like deque, dataframe, etc.), one would need the state downstream some time or another. But again, it's completely discretionary. 

Again, since this is something that is a production-mandated feature; users that don't want to use Kafka as a source, but still want the SDF state downstream, or just want to start  SDFs from a start  state, can handle the `state` however they wish downstream. For us, tying it to the Kafka checkpoint metadata is the eventual goal.

martindurant replied:
I think I agree with all of this. It is OK, from my view, to have a node which outputs different sorts of things depending on an init argument.

chinmaychandak commented on streamz/core.py:
```diff
@@ -867,12 +867,16 @@ def __init__(self, upstream, func, start=no_default, returns_state=False,
         self.returns_state = returns_state
         # this is one of a few stream specific kwargs
         stream_name = kwargs.pop('stream_name', None)
+        self.sdf_checkpoint = kwargs.pop('sdf_checkpoint', False)
```
I can update the doc string for `stream_name` [here](https://streamz.readthedocs.io/en/latest/api.html?highlight=stream_name#streamz.Stream). 

As for `sdf_checkpoint`, or whatever we end up calling it (`with_state`, like Martin suggested), I will update the doc strings for individual functions [here](https://streamz.readthedocs.io/en/latest/collections-api.html#dataframes).

Does that sound good?

martindurant replied:
Please do

chinmaychandak commented on streamz/core.py:
```diff
@@ -1063,7 +1070,7 @@ def cb(self):
             L, self._buffer = self._buffer, []
             metadata, self.metadata_buffer = self.metadata_buffer, []
-            m = [m for l in metadata for m in l]
+            m = [m for ml in metadata for m in ml]
```
This was part of the checkpointing PR. This time around (because of flake8 updates), it was showing flake8 [errors](https://github.com/scipy/scipy/issues/8389) saying that the name of variable `l` is ambiguous. That's why I changed the name to `ml` instead.

martindurant commented:
All good from me, aside from the rename and docstring updates.

chinmaychandak commented:
@martindurant, @CJ-Wright I have changed the variable name to `with_state`, added/updated plenty of doc strings, and also added a `@flaky` marker to the pertinent tests. 

All tests are passing. However, the coverage seems to have gone down, and I can't seem to understand why! The only changes made since the last commit (b5df231, that actually showed to have increased the code coverage) are the ones I mentioned above, and they shouldn't have impacted coverage.

martindurant commented:
The following diffs don't appear to be covered:
core.py:L879 return self._emit((self.state, x), metadata=metadata)
dataframe/core.py:L552 return Window(self.root.reset_index(), n=self.n, value=self.value,

martindurant commented:
@CJ-Wright , did you want another look at this?

chinmaychandak commented:
> The following diffs don't appear to be covered:
> core.py:L879 return self._emit((self.state, x), metadata=metadata)
> dataframe/core.py:L552 return Window(self.root.reset_index(), n=self.n, value=self.value,

@martindurant, I fixed these. But codecov still seems to complain. :(

martindurant commented on streamz/dataframe/core.py:
```diff
@@ -549,8 +549,7 @@ def example(self):
     def reset_index(self):
-        return Window(self.root.reset_index(), n=self.n, value=self.value,
-                      with_state=self.with_state, start=self.start)
+        return Window(self.root.reset_index(), n=self.n, value=self.value)
```
Why did this change back, was it intentional?

chinmaychandak replied:
Yes, I guess if one uses `reset_index` on a windowed-SDF operation, they would probably ignore the `start` state downstream anyways, so I removed it.

martindurant replied:
I don't mind which way this goes, but I would maybe have erred on including them in the method signature 
```python
def reset_index(self, with_state=None, start=None):
    return Window(self.root.reset_index(), n=self.n, value=self.value
        with_state=with_state if with_state is not None else self.with_state, 
        start=start or self.start)
```

chinmaychandak replied:
I don't know how I feel about this. Having a `start` in `sdf.window(...).reset_index(start=start)` can lead to unnecessary complications downstream. One should just use the appropriate `start` state in `window()` itself, rather than in `reset_index()`. This would conform the pandas/cudf `reset_index()` API.

martindurant replied:
OK :)
If ever it comes up that someone wants to do this, it will certainly be a corner case.

martindurant commented:
OK, I'm going to ignore that. I'll give @CJ-Wright a chance to say something yet (24hr), and then merge. 
For future reference, we could improve the coverage slightly by removing old py2 branches, which won't work now anyway.

chinmaychandak commented:
@martindurant Is this good to merge now?

> OK, I'm going to ignore that.

Any idea why it is complaining? This can become a problem for future PRs as well?
